src/geometry.cpp

```diff
@@ -0,0 +1,496 @@
+        << coord.lattice_x << "," << coord.lattice_y << ","
+        << coord.lattice_z << "). r=" << p->r();
+    write_message(msg, 1);
+  }
+
+  // Set the lattice indices.
+  coord.lattice_x += boundary.lattice_translation[0];
+  coord.lattice_y += boundary.lattice_translation[1];
+  coord.lattice_z += boundary.lattice_translation[2];
+  std::array<int, 3> i_xyz {coord.lattice_x, coord.lattice_y, coord.lattice_z};
+
+  // Set the new coordinate position.
+  const auto& upper_coord {p->coord_[p->n_coord_ - 2]};
+  const auto& cell {model::cells[upper_coord.cell]};
+  Position r = upper_coord.r;
+  r -= cell->translation_;
+  if (!cell->rotation_.empty()) {
+    r = r.rotate(cell->rotation_);
+  }
+  p->r_local() = lat.get_local_position(r, i_xyz);
+
+  if (!lat.are_valid_indices(i_xyz)) {
+    // The particle is outside the lattice.  Search for it from the base coords.
+    p->n_coord_ = 1;
+    bool found = find_cell(p, 0);
+    if (!found && p->alive_) {
+      std::stringstream err_msg;
+      err_msg << "Could not locate particle " << p->id_
+              << " after crossing a lattice boundary";
+      p->mark_as_lost(err_msg);
+    }
+
+  } else {
+    // Find cell in next lattice element.
+    p->coord_[p->n_coord_-1].universe = lat[i_xyz];
+    bool found = find_cell(p, 0);
+
+    if (!found) {
+      // A particle crossing the corner of a lattice tile may not be found.  In
+      // this case, search for it from the base coords.
+      p->n_coord_ = 1;
+      bool found = find_cell(p, 0);
+      if (!found && p->alive_) {
+        std::stringstream err_msg;
+        err_msg << "Could not locate particle " << p->id_
+                << " after crossing a lattice boundary";
+        p->mark_as_lost(err_msg);
+      }
+    }
+  }
+}
+
+//==============================================================================
+
+BoundaryInfo distance_to_boundary(Particle* p)
+{
+  BoundaryInfo info;
+  double d_lat = INFINITY;
+  double d_surf = INFINITY;
+  int32_t level_surf_cross;
+  std::array<int, 3> level_lat_trans {};
+
+  // Loop over each coordinate level.
+  for (int i = 0; i < p->n_coord_; i++) {
+    const auto& coord {p->coord_[i]};
+    Position r {coord.r};
+    Direction u {coord.u};
+    Cell& c {*model::cells[coord.cell]};
+
+    // Find the oncoming surface in this cell and the distance to it.
+    auto surface_distance = c.distance(r, u, p->surface_);
+    d_surf = surface_distance.first;
+    level_surf_cross = surface_distance.second;
+
+    // Find the distance to the next lattice tile crossing.
+    if (coord.lattice != C_NONE) {
+      auto& lat {*model::lattices[coord.lattice]};
+      std::array<int, 3> i_xyz {coord.lattice_x, coord.lattice_y, coord.lattice_z};
+      //TODO: refactor so both lattice use the same position argument (which
+      //also means the lat.type attribute can be removed)
+      std::pair<double, std::array<int, 3>> lattice_distance;
+      switch (lat.type_) {
```

Comment:
Might want to log that no switch case was hit in the default case if it is unexpected.